Adding index for ScalarType and LayoutType in DispatchKeyExtractor

charliechen0401 opened the pull request:
Summary:
As title,

Right now, I just search for index of IntType.
Will easily update to ScalarType and LayoutType when I am ready.

Test Plan: Added unit tests

Differential Revision: D19198534

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -102,15 +127,24 @@ struct DispatchKeyExtractor final {
   size_t num_args_;
+  // this is caching the index for ScalarType and LayoutType for args inside
+  // function schema.
+  c10::optional<size_t> scalar_type_index_;
```
We only want to consider ScalarType for dispatch if all TensorOptions arguments are present and in the correct order, right? So only then, this optional should be set. And that also means that you don't need to store `layout_index_`, because you can calculate it by offset from scalar_type_index_.

charliechen0401 replied:
yes. I forget that. I will just use one index here.

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -65,7 +65,12 @@ namespace detail {
-    return DispatchKeyExtractor(schema.arguments().size());
+    return DispatchKeyExtractor(
+      schema.arguments().size(),
+      // TODO: replacing intType with scalarType and LayoutType after
```
This must be fixed before we actually use this index anywhere.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -149,6 +149,10 @@ class CAFFE2_API OperatorHandle final {
     return operatorIterator_->op.options();
   }
 
+  const DispatchTable& dispatchTable() const {
```
Hm this exposes an implementation detail. Is there a way around this, e.g. friending the test or making it a unit test that doesn't use the Dispatcher but only the KeyExtractor?

smessmer reviewed (changes requested):
Requesting changes because the dual index to single index change will require some changes that I want to look at again

Commit:
Adding index for ScalarType and LayoutType in DispatchKeyExtractor (#31519)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31519

As title,

Right now, I just search for index of IntType.
Will easily update to ScalarType and LayoutType when I am ready.

Test Plan: Added unit tests

Differential Revision: D19198534

fbshipit-source-id: add6ef0925b652a5336bb37649617fec9b16e717

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -93,7 +97,23 @@ struct DispatchKeyExtractor final {
+
 private:
+  static c10::optional<size_t> getOptionalIndexForArgType_(
+    TypePtr type,
```
`const TypePtr&` to avoid the refcount bump is a bit faster, TypePtr is a shared_ptr

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -65,7 +65,11 @@ namespace detail {
+      schema.arguments().size(),
+      // TODO: replacing intType with scalarType and LayoutType after
+      // both type are defined correctly in jit_type.
+      getOptionalIndexForArgType_(IntType::get(), schema.arguments()));
```
Note that in the final version (i.e. once it uses ScalarType instead of int), just looking for ScalarType is not enough. You also need to check that the other arguments (Layout, Device) are present and in the right position.

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -102,15 +122,21 @@ struct DispatchKeyExtractor final {
   // fallthrough
   // TODO: a potential optimization is to store a bitfield of arg locations,
   size_t num_args_;
+  // this is caching the index for ScalarType for args inside
```
Maybe, in the final version also add a bit more comment here, explaining that if this is set, then the argument at this index must be ScalarType, at index+1 and index+2 must be ... and if this is not set, then we know that the op does not have this constellation of args

smessmer commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -39,6 +39,29 @@ struct MockKernel final : OperatorKernel {
+  EXPECT_EQ(scalarTypeIndex.value(), 2);
+}
+
+TEST(OperatorRegistrationTest, whenRegisterOperatorWithMultipleArgToBeIntType_thenPassAndReturnFirstIndex) {
```
Once we change this to ScalarType, it is not "the first ScalarType" we want but the one that is in the correct constellation with DeviceType and Layout. And if we have multiple of those triplets, we should probably throw an error rather than taking the first one because we don't support that.

smessmer reviewed (approved):
Hm looking at all the "once we use ScalarType instead of int" comments, I'm not sure it makes sense to land this before we changed it to ScalarType. This by itself doesn't unblock anything or add any features, it just adds a hack without actually needing or using it in this PR. Up to you though but if you do land it, please address all the comments in the PR where you do change it to ScalarType.

charliechen0401 commented:
Thanks for the comment @smessmer . I feel now I will just keep this PR not land till next year(when we will have a discussion on how to get jit types steps by steps). 
Before that, I will expend the diff more to resolve the comments + including getDispatchKeyBoxed related changes.

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 5a898f9b:


* **1/1** broken upstream at merge base df9d5b8a (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=df9d5b8a7772010f4faa0f18a798634ac9ba7635&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=5a898f9b0a34b379bce1da3275733f494b313663).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_clang5_asan_test](https://dr.pytorch.org/build-details.html?build_id=78526563)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>
